runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler.cpp

```diff
@@ -0,0 +1,156 @@
+#include "VMHelpers.hpp"
+#include "UnsafeAPI.hpp"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+/**
+ * openj9.internal.foreign.abi.InternalUpcallHandler: private native long allocateUpcallStub(UpcallMHMetaData mhMetaData, String[] cSignatureStrs);
+ *
+ * @brief Request the JIT to generate an upcall thunk of the specified java method
+ *
+ * @param mhMetaData[in] a data object that consists of the upcall handler and the metadata for MH resolution
+ * @param cSignatureStrs[in] An array of the native signature string for the requested java method during the upcall
+ * @return the address of the native thunk generated by JIT.
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler_allocateUpcallStub(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9UpcallMetaData *upcallMetaData = NULL;
+	j9object_t mhMetaData = NULL;
+	J9UpcallNativeSignature *nativeSig = NULL;
+	J9UpcallSigType *sigArray = NULL;
+	void *thunkAddr = NULL;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	j9object_t cSigStrs = (j9object_t)currentThread->sp[0];
+	/* the last element of the array is the signature of return type */
+	U_32 sigCount  = J9INDEXABLEOBJECT_SIZE(currentThread, cSigStrs);
+
+	/* Note: the J9UpcallMetaData pointer will be stored in the generated thunk as data
+	 * in which case it is released only when the generated thunk memory is released
+	 */
+	upcallMetaData = (J9UpcallMetaData *)j9mem_allocate_memory(sizeof(J9UpcallMetaData), J9MEM_CATEGORY_VM_FFI);
+	if (NULL == upcallMetaData) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+
+	nativeSig = (J9UpcallNativeSignature *)j9mem_allocate_memory(sizeof(J9UpcallNativeSignature), J9MEM_CATEGORY_VM_FFI);
+	if (NULL == nativeSig) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	sigArray = (J9UpcallSigType *)j9mem_allocate_memory(sizeof(J9UpcallSigType) * sigCount, J9MEM_CATEGORY_VM_FFI);
+	if (NULL == sigArray) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto freeAllMemoryThenExit;
+	}
+
+	for (U_32 sigIndex = 0; sigIndex < sigCount; sigIndex++) {
+		j9object_t sigStrObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, cSigStrs, sigIndex);
+		char sigBuffer[J9VM_NATIVE_SIGNATURE_STRING_LENGTH] = {0};
+		/* The simplified signature string in cSig for parameter/return type is converted at java level.
+		 * e.g.
+		 * "4#I" represents a 4-byte integer
+		 * "20#[5:I]" represents a 20-byte struct for {int a[5]}
+		 * "16#[I(4)J]" represents a 16-byte struct for {int, padding(4 bytes), long}
+		 *
+		 * Note: the last element is the signature string for the return type.
+		 */
+		char *cSig = copyStringToUTF8WithMemAlloc(currentThread, sigStrObject,
```

Comment:
Can this fail?

Reply:
Just added the null check for `cSig` in this case.

Reply:
OK, please re-check all of this code for other fail/OOM situations as this not the first that was missing the check.

Reply:
I just went through the code in `OutOfLineINL_openj9_internal_foreign_abi_InternalUpcallHandler.cpp` and confirm that a null check is done for each memory related pointer.